sql/src/main/java/io/crate/executor/task/LocalMergeTask.java

```diff
@@ -138,40 +146,41 @@ public void onFailure(@Nonnull Throwable t) {
                 result.setException(t);
             }
         });
-
-        for (final ListenableFuture<TaskResult> upstreamResult : upstreamResults) {
-            Futures.addCallback(upstreamResult, new FutureCallback<TaskResult>() {
-                @Override
-                public void onSuccess(@Nullable TaskResult rows) {
-                    assert rows != null;
-                    traceLogResult(rows);
-                    boolean shouldContinue;
-
-                    try {
-                        shouldContinue = mergeOperation.addRows(rows.rows());
-
-                        if (countdown.decrementAndGet() == 0 || !shouldContinue) {
-                            mergeOperation.finished();
-                        }
-                    } catch (Throwable ex) {
-                        onFailure(ex);
-                        logger.error("Failed to add rows", ex);
+        // transform taskresults into bucket futures
+        BucketPage page = new BucketPage(Lists.transform(upstreamResults, new Function<ListenableFuture<TaskResult>, ListenableFuture<Bucket>>() {
+            @Nullable
+            @Override
+            public ListenableFuture<Bucket> apply(@Nullable ListenableFuture<TaskResult> future) {
+                assert future != null : "taskresult future is null";
+                return Futures.transform(future, new Function<TaskResult, Bucket>() {
+                    @Nullable
+                    @Override
+                    public Bucket apply(@Nullable TaskResult taskResult) {
+                        assert taskResult != null : "taskresult is null";
+                        traceLogResult(taskResult);
+                        return taskResult.rows();
                     }
-                }
-
-                @Override
-                public void onFailure(@Nonnull Throwable t) {
-                    ramAccountingContext.close();
-                    statsTables.operationFinished(operationId, Exceptions.messageOf(t),
-                            ramAccountingContext.totalBytes());
-                    result.setException(t);
-                }
-            }, threadPool.executor(ThreadPool.Names.GENERIC));
-        }
+                });
+            }
+        }));
+        mergeOperation.pageDownstream().nextPage(page, new PageConsumeListener() {
+            @Override
+            public void needMore() {
+                // currently only one page
+                LOGGER.trace("{} need more", mergeNode.contextId());
+                finish();
```

Comment:
if the first page contained less rows than expected, this method is called, so its a valid path. we finish nonetheless, as there is nothing left to come.